Tic Tac Toe - first view

SergoShe opened the pull request:
Необходимо реализовать игру "Крестики-нолики". Принцип следующий.
На входе в игру есть меню, в котором можно запустить игру или выйти. При запуске игры надо выбрать количество игроков (1 или 2). После этого запускается игра. Игра ведётся в текстовом виде:
Turn: Player 2
Current:
Х * *
* О *
Х * *

В качестве хода принимается клетка в которую нужно указать (например 2,1 - вторая строка первый столбец)

По завершению показываются результаты серии игр и предлагается начать новую/ закончить серию. При окончании серии - возврат в главное меню.

При выборе режима 1 игрок, необходимо выбрать сложность ИИ.
easy - бот ставит метку случайным образом в любую свободную
hard - бот ставит метку максимально правильным образом

SergoShe committed:
Tic Tac Toe - first view

SergoShe committed:
some changes and package distribution

rs-pluss commented on src/gameplay/Gameplay.java:
```diff
@@ -0,0 +1,119 @@
+        return position;
+    }
+
+    private void init() {
```
У тебя есть отдельный класс инициализатор, правильнее будет в нём собирать все данные а сюда передавать уже готовые и ничего не инициализировать - здесь уже процесс игры идёт

rs-pluss commented on src/gameplay/Score.java:
```diff
@@ -0,0 +1,28 @@
+package gameplay;
+
+public class Score {
+    private int scorePlayer1;
```
Ты нигде не инициализируешь эти переменные, понимаешь почему они не выдают null?

rs-pluss commented on src/gameplay/difficulty/AIHard.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.ArrayList;
+import java.util.stream.IntStream;
+
+public class AIHard extends Player {
```
Правильнее будет сделать иерархию в который AIHard и AIEasy будут унаследованы от общего класса.

rs-pluss commented on src/gameplay/difficulty/Player.java:
```diff
@@ -0,0 +1,42 @@
+package gameplay.difficulty;
```
У тебя игрок в пакете "Сложность", это не так

rs-pluss commented on src/initializer/PlayerDifficulty.java:
```diff
@@ -0,0 +1,7 @@
+package initializer;
+
+public enum PlayerDifficulty {
```
Человек, это не Сложность :)

SergoShe committed:
some changes and fixes

rs-pluss commented on src/initializer/TicTacToeInitializer.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    private Player selectNumberOfPlayers() throws IOException {
```
Название метода не соответствует возвращаемому значению

rs-pluss commented on src/initializer/TicTacToeInitializer.java:
```diff
@@ -0,0 +1,76 @@
+public class TicTacToeInitializer {
+    BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+
+    public void start() {
```
Какой смысл подобного оборачивания методов?

rs-pluss commented on src/initializer/TicTacToeInitializer.java:
```diff
@@ -0,0 +1,76 @@
+        work();
+    }
+
+    private void work() {
```
Initialize

rs-pluss commented on src/initializer/player/bot/AIHard.java:
```diff
@@ -0,0 +1,92 @@
+        }
+    }
+
+    private int[] findFreeCells(Sign[][] table) {
```
Где то я это уже видел

rs-pluss commented on src/gameplay/Gameplay.java:
```diff
@@ -0,0 +1,99 @@
+                    currentPlayer = changePlayer(currentPlayer);
+                }
+            }
+            score.showScore();
```
Всё что ниже не должно относиться к этому методу, а лучше - к этому классу

rs-pluss commented on src/gameplay/Gameplay.java:
```diff
@@ -0,0 +1,99 @@
+            }
+            score.showScore();
+            boolean isContinue = false;
+            while (!isContinue) {
```
Переменная "Продолжить = false"
И в цикле  (до тех пор пока  НЕ продолжать)

rs-pluss commented on src/gameplay/Gameplay.java:
```diff
@@ -0,0 +1,99 @@
+        Player currentPlayer = firstPlayer;
+        boolean isEndGame = false;
+        while (!isEndGame) {
+            board.initTable();
```
Это плохой паттерн. Не надо использовать много раз один и тот же объект.

Расскажи мне свои предположения почему.

rs-pluss commented on src/gameplay/Board.java:
```diff
@@ -0,0 +1,72 @@
+
+    }
+
+    public void initTable() {
```
Это надо уносить в конструктор

SergoShe committed:
some changes

rs-pluss commented on src/session/GameMode.java:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    public static GameMode valueOf(int value) {
```
Весь этот код повторяется, есть идеи как переиспользовать?

rs-pluss commented on src/session/Score.java:
```diff
@@ -0,0 +1,35 @@
+    private int secondPlayerScore = 0;
+    private int drawScore = 0;
+
+    /*public void incrementFirstPlayer() {
```
?

rs-pluss commented on src/session/Winner.java:
```diff
@@ -0,0 +1,8 @@
+    DRAW,
+    FIRST_PLAYER,
+    SECOND_PLAYER,
+    UNKNOWN
```
Вообще с точки зрения чистоты кода fallback - вариант везде и всегда пишется первым

rs-pluss commented on src/session/initializer/NumberPlayers.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public enum NumberPlayers {
```
of

rs-pluss commented on src/session/gameplay/Board.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public void initTable() {
+        Arrays.stream(table).forEach(x -> Arrays.fill(x, Sign.SIGN_EMPTY));
```
Зачем это?

rs-pluss commented on src/session/gameplay/Board.java:
```diff
@@ -0,0 +1,74 @@
+        return table[2][0] == table[1][1] && table[1][1] == table[0][2];
+    }
+
+    private String printSign(int row, int column) {
```
Метод ничего не выводит в консоль

rs-pluss commented on src/session/initializer/Initializer.java:
```diff
@@ -0,0 +1,61 @@
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+public class Initializer {
```
обычно подписывают initializer "чего"

rs-pluss commented on src/session/player/bot/AIHard.java:
```diff
@@ -0,0 +1,81 @@
+        ArrayList<Integer> bestMoves = new ArrayList<>();
+        for (int freeCell : findFreeCells(board.getTable())) {
+            Position currentPos = new Position(freeCell);
+            board.setSign(currentPos, sign);
```
Вот так нельзя делать, ты используешь общую доску и выставляешь в ней знак.
Через строку ты этот знак убираешь, но, при развитии сервиса эти строки могут разъединиться, и это может привести к фатальным ошибкам. Если бы ты разрабатывал вместо tic-tac какой нибудь банковский сервив, подобный код может привести к милионным потерям денег.

rs-pluss replied:
Аналогичная проблема может быть когда столкнёшься с многопоточными программами - там временное изменение объекта по общей ссылке тоже может привести к фаталу.

rs-pluss commented on src/session/Session.java:
```diff
@@ -0,0 +1,64 @@
+
+    private void initPlayers() {
+        init.initialize();
+        firstPlayer = init.getFirstPlayer();
```
А где ротация игроков?

rs-pluss commented on src/session/Session.java:
```diff
@@ -0,0 +1,64 @@
+    private void initPlayers() {
+        init.initialize();
+        firstPlayer = init.getFirstPlayer();
+        secondPlayer = init.getSecondPlayer();
```
По хорошему игроки во время сессии остаются те же самые, тебе в конструкторе положить их в этот класс а потом уже не изменять

SergoShe committed:
- Корректировка алгоритма AIHard
- Скорректированы названия методов и классов
- Убраны лишние комментарии и методы

rs-pluss commented on src/session/player/bot/AIHard.java:
```diff
@@ -0,0 +1,81 @@
+        ArrayList<Integer> bestMoves = new ArrayList<>();
+        for (int freeCell : findFreeCells(board.getTable())) {
+            Position currentPos = new Position(freeCell);
+            Board boardClone = board.clone();
```
Правильнее будет обернуть это всё в отдельны метод - в текущем только клонировать доску. Таким образом у тебя будет явное разделение

SergoShe committed:
- Небольшое изменение в AIHard

SergoShe committed:
- Небольшое изменение
